### ISSUE TITLE
Implement AMM Curve For Like-Asset Pairs In Agoric

### ISSUE THREAD

**Issue by danishyasin33** [robor-systems/agoric-amm-curve#1]
### Description
Implement a curve for Agoric's Automated Market Maker (AMM) which minimizes slippage, similar to [curve.fi's](https://curve.fi/) StableSwap in Ethereum.

Completed work will include:

A new version of bondingCurves.js using new curve structure
An update to Agoric's MultiPool Autoswap contract to include a term that chooses which bonding curve to use

### Context
Standard X*Y=K automated market maker (AMM) curves are meant for asset pairs which have relative volatility and require price responses (slippage) based on trades.

However, there is a large demand for AMM swaps of like-assets (e.g., USDC-USDT, WBTC-TBTC, ETH-sETH) - particularly stablecoin pairs - for which the slippage driven by X*Y=K curves is not desirable. [Curve.Fi's](https://curve.fi/) StableSwap implementation on Ethereum has had great success driving high volume of these trades.

Agoric's stable local currency RUN will need to trade against other stablecoins with minimal slippage. This adjusted curve implementation will allow for that.

### Acceptance Criteria
Price slippage for a given trade and pool size is comparable to StableSwap
Pool value never declines on swap, as confirmed by testing framework
Rounding of values gives preferential treatment to the pool
Bonding curve price calculations use ratios and amountMath rather than Nat
Offer safety and rights conservation are not violated on trades

### References
- Agoric Documentation: https://agoric.com/documentation/getting-started/
- Curve.Fi's StableSwap implementation: https://curve.fi/files/stableswap-paper.pdf
- Agoric's AMM X*Y=K curve logic: https://github.com/Agoric/agoric-sdk/blob/master/packages/zoe/src/contractSupport/bondingCurves.js
- Agoric's AMM: https://github.com/Agoric/agoric-sdk/tree/master/packages/zoe/src/contracts/multipoolAutoswap

**Comment by Hussain-7** [comment]
## Update 31st Jan 2022
- Read the ticket. Understood every single term mentioned in the ticket.
- Got the main idea of the task that is to implement a curve for `agoric stable coin` just like what is implemented for curvefi essentially to `reduce slippage` which is the difference between the price you expect to get on the crypto you have ordered and the price you actually get when the order executes.
- Revised my concept of the `constant product amm equation` and how it is used to set price for assets in a liquidity pools.(reviewed amm again just for quick recap).
- Understood the `acceptance criteria`.
- Read the curvefi white paper to understand the implementation of stable swap curvefi is using.Which is somewhat what we need to use for agoric.Just have to figure out the equation after some research. Still have to finish reading this paper.
-Overviewed the current implementation of `bonding curves` in `agoric sdk` which uses the `contant product equation` that should not particularly used for the stable coins. Hence that is for me to figure out the equation to use.
### Todo's next:
- Have to complete reading the white paper.
- Develop complete understanding of the bondingCurve.js file in agoric-sdk with respect to code.
- Understand the `AutoSwap` contract so that I'm able to update it when required.
- Will probably have a know how of how to implement this new equation after reading through the white paper and some more documentation related to curvefi's implementation.
- Then will begin implementation if I find some idea.

**Comment by Hussain-7** [comment]
## Update 1st Feb 2022
- First understood a bit of `defi infrastructure`. That made many thing mentioned in the white paper more clearer.
- The equation we are required to implement is as follows:
  ![image](https://user-images.githubusercontent.com/55734494/152237386-a61fe9da-c86c-4ba8-a534-d45a1447cd11.png)
- Spent some time to derive this equation to understand the working that would be required to get this equation and use it to find `exchange amounts` of two stable assets in a `amm`. Since some steps were missing in the white papers had to solve from the start and derive the end equation.
- Also understood it graphically which made everything very clear.
- Basically it uses combines the concept of `constant product equation` which `uniswap` uses and the constant sum equation. Which are as follows - X*Y=(D/2)^2 and X+Y=D 
- Will add a `detailed derivation` with all steps later. That actually explains most of the concept behind the equation above.
- **Effort Required:**
  - I think this task should be completed by the end this week if not at max would require one additional day.
- Most of the part is already clear will start working on it from today.

**Comment by Hussain-7** [comment]
## Update 2nd Feb 2022
- First tested out the already written `bondingcurve` functions by running unit tests using `ava cli`.
  - Tested the `getInputPrice()` and `getOutputPrice()`. 
  - `getInputPrice` : This function contains logic for calculating how much should be given back to the user in exchange for what they sent in.
  - `getOutputPrice()` :  This function contains logic for calculating how much should be taken from the user in exchange for what they want to obtain.
- There is also concept of `fee cut` on the input value when given for exchange with some other asset. 
- The current `uniswap` implementation was explained in a link mentioned in the code. So cleared some concepts from there. 
  - Reference : https://github.com/runtimeverification/verified-smart-contracts/blob/uniswap/uniswap/x-y-k.pdf
- Had to clear some `basic conversion` concepts by running test and testing the already available functions.
- After that created a new `bondingCurve` file and added a new function for `getInputPrice()` which uses the equation mentioned above. Some issue are still to be resolved.
- Also was testing this function using ave cli and unit tests.

**Comment by Hussain-7** [comment]
## Update 3rd Feb 2022
- Implemented the `getInputPrice()` using the following two formula's: 

  1. <img src="https://user-images.githubusercontent.com/55734494/152454662-c6a2c176-c768-4611-bd03-d1d869b1d6f6.jpeg" data-canonical-src="https://user-images.githubusercontent.com/55734494/152454662-c6a2c176-c768-4611-bd03-d1d869b1d6f6.jpeg" width="350" height="150" />
  2. <img src="https://user-images.githubusercontent.com/55734494/152454673-9183522f-f45b-4db1-9473-ed7610e37639.jpeg" data-canonical-src="https://user-images.githubusercontent.com/55734494/152454673-9183522f-f45b-4db1-9473-ed7610e37639.jpeg" width="350" height="150" />
### Step Required :
1. Decide on the amplification coefficient  A. Most optimal value for A was given in the white paper so used that. 
2. Calculate D. The formula for D as shown above is derived from the same formula mentioned in white paper. To get value of D it needs to be solved iteratively.
4. Then we solve the equation for y, given that the parameters A, D and x are all known.
### Progress 
- Have implemented the above working in the `getInputPrice()` and it is returning the correct values for a certain range.
- For the case where difference between the amount of assets x and y in the liquidity pool is `less` than `100times`, we can successfully find the amount of asset to return and the price.
- For the case where difference in amounts of `swap in` and `out` asset in the liquidity pool is `more` than `100times` then the formulas above donot converge to a solution and give error due to negative values.
- Will need to find the solution for the error case.
- Found out a stable swap asset simulator where I can choose value for A,X,Y and swap amount. Then using that we can get the amount of asset returned and the price. Reference : https://miguelmota.com/blog/understanding-stableswap-curve/
- Once the issue with the case mentioned above is resolved can move to converting every value to amounts instead of Nat or bigInts.

**Comment by Hussain-7** [comment]
## Update 4th Feb 2022
- As mentioned in the previous update that there were some cases in which value returned is wrong.
- To solve that I even found a `stableswap Math` library in `Rust`. Went through all the functions and  implemented almost the same logic. 
   - Reference : https://docs.rs/stable-swap-math/latest/i686-pc-windows-msvc/src/stable_swap_math/curve.rs.html
- But Still that also gave me the same issue for the mentioned cases.
- Figured the `issue` is with the `amplification constant` which I'm keeping as constant and not changing during calculation.
- Can't find any solution which is different from what I am doing. 
- So currently I have implemented the `getInputPrice()` methods with two different ways. One which I derived myself and the other one which is used in the `stableSwap math` library in Rust.

**Comment by Hussain-7** [comment]
## Update 7th Feb 2022
### Progress
- Implemented `getInputPrice()` function and all the functions it requires with a different approach.
- Previously I was solving equations using `fixed point iteration` method that `converges` slowly.
- Now I have implemented the `newtons method` to solve the equation which best choice method for this scenario. Mostly in where ever I referenced , this method is preferred method for solving linear equations.   
- As mentioned previously was facing issues with certain ranges of input. That resulted to negative or value of swap to asset amount.
- Now we never get a value below 0. 
### Next todos
- Will Add the `basis_point` conversions to the input values (liquidities and swap in amount).
- After that will work on getOutputPrice().

**Comment by Hussain-7** [comment]
## Update 8th Feb 2022
### Progress
- Implemented the `getOutputPrice()` function
- Wrote some more tests to validate this function.
- Added `basis_point` conversion throughout the implementation so that we can decimal weightage in `bigint` as well.
- Also implemented `price` calculation for both `getInputPrice()` and `getOutputPrice()` function
- Thoroughly tested out the results by comparing with output provided by the `reference stable swap graph`. 
- Worked on converting all the calculations to `Amounts`. Still some conversions left.
### Next todos
- Minor fix required in price calculation for output price function.
- Have to complete conversion of all `Nat` value being used to AmountMath `amounts`
- Run test to check out everything with amounts is working fine or not.
- Everything should be done by tomorrow.

**Comment by Chris-Hibbert** [comment]
Hi, I'm Chris Hibbert, one of the developers at Agoric. 

Here's our current [AMM implementation](https://github.com/Agoric/agoric-sdk/tree/master/packages/run-protocol/src/vpool-xyk-amm).  It doesn't follow the uniSwap model as closely and doesn't use `bonding-curve.js`. It's documented [here](https://github.com/Agoric/agoric-sdk/tree/master/packages/run-protocol/src/vpool-xyk-amm/constantProduct#readme).  I don't think you are responsible for integrating into this framework. I think a stable curve could be integrated in as a virtual pool.

Can I read your code in progress?

It appears that the Rust code is linearly scaling `A` over time, which doesn't seem to match either `stableswap-paper.pdf` or the "Understanding Stableswap" paper. (see `compute_amp_factor` on line 142.) Which model are you following?

Are you implementing a two-currency version, or the generalized n-currency solution? 

The stableswap paper doesn't talk about liquidity tokens. What is your approach based on?

**Comment by Hussain-7** [comment]
- Thank you for clearing out the previous Amm implementation, according to the initial description: https://github.com/robor-systems/agoric-amm-curve/issues/1#issue-1119434076, amm's bonding curve logic is in the bondingCurve.js file so we assumed it was used in the amm for trade aswell.
- As you mentioned that the stable curve could be integrated in as a `virtual pool`. So, we will deliver a custom contract with the new stable curve calculation. We believe this should be the easiest way to test it in a virtual environment. Would that suffice? 
- You can check out the stable curve code here: [Stable bonding curve Implementation](https://github.com/robor-systems/agoric-amm-curve/blob/bonding-curve-update/packages/zoe/src/contractSupport/stableBondingCurves.js)
- And we've also created tests that you can run to check out the functionality: [Stable bonding curve tests](https://github.com/robor-systems/agoric-amm-curve/blob/bonding-curve-update/packages/zoe/test/unitTests/contractSupport/test-stableBondingCurves.js)

### Model used:
- I have implemented a generalized n-currency solution.
- Please ignore the Rust library. It was referenced above just to see which method to use to solve the non-linear equations. Their equation used was not followed anywhere in the code. 
- Im using the the equations provided in `stableswap-paper.pdf`. This was the final equation:
  <p align="center">
  <img src="https://user-images.githubusercontent.com/55734494/153146549-917c782c-4664-4e57-a8ee-67cb730bb5db.png" />
</p>

-  So the unknown in the above equation are only D and y. Our aim is to actually to find y which is the amount of coin y left in portfolio of Coins after the increase in amount x due to swap in of coin x.
-  Then using `Newton Raphson method` for solving non-linear equations I have solved the above equation iteratively.



- Apologies for the confusion but by `liquidity of a coin` (where ever mentioned in above tickets) I meant the amount of a certain coin in the `portfolio of coins` provided to the `stable swap curve` function.
- Feel free to check out the tests mentioned above and guide us if any changes are required.

### In progress
- The conversion of all value used to AmountMath amounts is still is progress , apart from that if you find any issue please do tell

**Comment by danishyasin33** [comment]
Hi @Chris-Hibbert, thank you for your feedback, if our approach is still unclear after reading Hussain's comment then let's have a call to discuss this task so that we can get a better idea of how to approach this.

**Comment by Chris-Hibbert** [comment]
`t.deepEqual(output, output)` [Lines 24 & 39]  seems like an unhelpful test.  I think both should be `t.deepEqual(output, expectedOutput);`.

The approach looks sound. When I review this, I'll want more comments on the roles of `b`, `c`, and `s` in `getY()`.   Can we characterize when `getY()` would be expected not to converge?

I could also use more help understanding the iterated expression in `getD()`:  What is the derivation (sorry!) of the derivative expression used in newton's recursion? Why do we believe that it should converge? (Is it sufficient that it's continuous? I certainly grant that.)

I like the approach taken to charging fees.

**Comment by Hussain-7** [comment]
## Update 9th Feb 2022
- Thanks for giving the feedback again. Will leave an update soon on the concepts you have asked to clear.
- And yes will add more comments in code more for better explaining of each and every variable.
### Progress:
- Resolved some issues in `output price calculation`.
- Wrote some test for output price functions and compared it to results available on web.
- Read details on how the `amm implementation` calculates prices .It uses the ratios.
- So have converted all the `price calculations` in stableBondingCurve.js to `amount` and `Ratios` instead of `nat vales`.
- Read the constant Product AMM description provided in docs.
### Next up:
- Create a contract in amm. To shift all the code for stable bonding curve to amm's contract.
- Will complete the contract and call its functions from the frontend to test out.
- Next will leave a detailed update on the some more things that chris have asked. 
- Add up comments after every calculation to explain the formula being  used.

**Comment by Hussain-7** [comment]
### For Convergence of Both D and Y:
- We can find a root of a `function or equation` being solved iteratively using `newton raphson` method as long as the following conditions satisfy:
<p align="center">
<img src="https://user-images.githubusercontent.com/55734494/153313061-24c73164-61f2-44c7-a789-09d4359271be.png" data-canonical-src="https://user-images.githubusercontent.com/55734494/153313061-24c73164-61f2-44c7-a789-09d4359271be.png" width="350" height="276" />
</p>

### Function for D:
- By rearranging the `stable swap equation` and bringing all coefficients of D on one side of the equation to make f(D)=0 equation. Where f(D) AND f`(D) are as follows:

<p align="center">
<img src="https://user-images.githubusercontent.com/55734494/153314415-153bb2d0-1c34-4802-944a-213beecceaa4.jpeg"  width="350" height="250"/>
<p>

- So after getting a `simplified equation for Di+1` from the above `formula`, using limits to -ve and +ve infinity I tested out that `Di+1` is always less than 1 and will converge every time. 
- Same goes for the `equation for Y` as well. In similar manner equation for y was found and tested out for convergence.

### Some more details
- About the line `t.deepEqual(output, output)` .During testing, since I was trying multiple possible implementations. It was requiring me to change the expected output value repeatedly .Hence for the development time I changed it to output value, So at least I can call the functions and check output without tests failing.
- The line `throw new error('y does not converge')` was just added to test out a particular use case which no longer exist now hence that is also removed from the code.
- For `b`, `c`, and `s` in `getY()` , I have added comments about the equation these variable hold value for. And what the combined equation of y sums up to.
- Feel free to checkout the **latest branch** which is `bonding-curve-amount`.

**Comment by Hussain-7** [comment]
## Update 10th Feb 2022
### Progress:
- Revised all the `formula` and `concepts` for commenting in code and explaining the concept we used to figure out whether convergence on y and d is always true.
- New implementation in price calculation that is `central token` is the medium in conversion of every swap. That if `token A` is to be swap to token B. Token A is converted and to find equivalent Run and the Run is converted to equivalent `token B`.
- Created a stable swap contract which has this stable swap curve functionality.
- Have commented in code where ever possible.

**Comment by Hussain-7** [comment]
## Update 14th Feb 2022
### Progress in amm-curve:
- To write test needed to configure the `contract` folder workspace to run test for `stableSwapContract`.  
- Then wrote multiple test for both `getInputPrice()` and `getOutputPrice()`.
- Tested out results by:
  - Manually calculated and verified some results.  
  - Tested the get Input price function with the results on the web and the get input price function could be used to verify get output price function.
- Found a `bug` during testing, fixed that.
- Latest code has been pushed to the `agoric-amm`  repo;
### Progress agoric-card-store multi-dapp Issue:
- Had conversation in the agoric-channel and figured the problem is with the amount of run in our wallet hence it should be increased and to do that: 
  - We cannot just mint `Run` because it is not possible since we don't have run minter.
  - So the other way out is that we use the treasury to swap `BLD` with `RUN` to increase amount of run in our wallet.
  - Successfully did that but didn't tested out `multi-user-dapp` after that
- Will test out the multi user next now that we have increased run in our wallet.

**Comment by Hussain-7** [comment]
## Update 23th Feb 2022
- Reviewed all the issues mentioned in `Pr`
- Fixed all issues which were to be fixed.
- Explained some formulas and problems left.
- Wrote `unit tests` for `d` values as well.
- Latest code is pushed `stable-curve-amount` branch.
- Stable Swap Main calculation is summarized here : [Link](https://drive.google.com/file/d/1mTgRy_iYQJgmoeELwy97vjcGLvskBIDo/view?usp=sharing)
## Next Up
- Will try to finish up on cardstore issues.

**Comment by Hussain-7** [comment]
### Update 28th Feb 2022
- Mostly time spent on ratio issue which was causing `bigints` to get very large.
- Fixed that by refactoring the formula a bit.
- Created a `pdf guide` for formula calculation. [Linked here](https://drive.google.com/file/d/1mTgRy_iYQJgmoeELwy97vjcGLvskBIDo/view?usp=sharing)
- Did testing for extreme cases.
- Have created a new Pr for the latest code.
- Worked on card store issues with haseeb.
### Next Up
- Will work on card store  with haseeb and finish it up tommorrow

**Comment by danishyasin33** [comment]
No further action required
